### PR TITLE
[inductor] fix CppPythonBindingsCodeCache code build failed on Windows (long to int64_t).

### PR DESCRIPTION
**Pull request by xuhancn**
Test case:
```cmd
pytest test/dynamo/test_dynamic_shapes.py -k test_donated_buffer1_dynamic_shapes
```
Error msg:
```cmd
(win_mkl_static) D:\xu_git\dnnl_cb\pytorch>cl /I C:/Users/Xuhan/.conda/envs/win_mkl_static/Include /I C:/Users/Xuhan/.conda/envs/win_mkl_static/lib/site-packages/torch/include /I C:/Users/Xuhan/.conda/envs/win_mkl_static/lib/site-packages/torch/include/torch/csrc/api/include /I C:/Users/Xuhan/.conda/envs/win_mkl_static/lib/site-packages/torch/include/TH /I C:/Users/Xuhan/.conda/envs/win_mkl_static/lib/site-packages/torch/include/THC /D TORCH_INDUCTOR_CPP_WRAPPER /D C10_USING_CUSTOM_GENERATED_MACROS /D CPU_CAPABILITY_AVX512 /DLL /MD /O2 /std:c++20 /wd4819 /wd4251 /wd4244 /wd4267 /wd4275 /wd4018 /wd4190 /wd4624 /wd4067 /wd4068 /EHsc /openmp /openmp:experimental C:/Users/Xuhan/AppData/Local/Temp/torchinductor_Xuhan/p3/cp3eocnerhprcphyxo3d3aeh7kn6pguwgv6flnhuhavs5zhh3fi4.cpp /arch:AVX512 /LD /FeC:/Users/Xuhan/AppData/Local/Temp/torchinductor_Xuhan/p3/cp3eocnerhprcphyxo3d3aeh7kn6pguwgv6flnhuhavs5zhh3fi4.pyd /link /LIBPATH:C:/Users/Xuhan/.conda/envs/win_mkl_static/libs /LIBPATH:C:/Users/Xuhan/.conda/envs/win_mkl_static/lib/site-packages/torch/lib torch.lib torch_cpu.lib torch_python.lib sleef.lib c10.lib
Microsoft (R) C/C++ Optimizing Compiler Version 19.37.32825 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

cl : Command line warning D9025 : overriding '/openmp' with '/openmp:experimental'
cp3eocnerhprcphyxo3d3aeh7kn6pguwgv6flnhuhavs5zhh3fi4.cpp
C:/Users/Xuhan/AppData/Local/Temp/torchinductor_Xuhan/p3/cp3eocnerhprcphyxo3d3aeh7kn6pguwgv6flnhuhavs5zhh3fi4.cpp(19): warning C4849: OpenMP 'simdlen' clause ignored in 'simd' directive
C:/Users/Xuhan/AppData/Local/Temp/torchinductor_Xuhan/p3/cp3eocnerhprcphyxo3d3aeh7kn6pguwgv6flnhuhavs5zhh3fi4.cpp(55): error C2338: static_assert failed: 'arg type must be pointer or long'
C:/Users/Xuhan/AppData/Local/Temp/torchinductor_Xuhan/p3/cp3eocnerhprcphyxo3d3aeh7kn6pguwgv6flnhuhavs5zhh3fi4.cpp(79): note: see reference to function template instantiation 'T parse_arg<int64_t>(PyObject *,size_t)' being compiled
        with
        [
            T=int64_t
        ]
C:/Users/Xuhan/AppData/Local/Temp/torchinductor_Xuhan/p3/cp3eocnerhprcphyxo3d3aeh7kn6pguwgv6flnhuhavs5zhh3fi4.cpp(56): error C2440: 'static_cast': cannot convert from 'void *' to 'int64_t'
C:/Users/Xuhan/AppData/Local/Temp/torchinductor_Xuhan/p3/cp3eocnerhprcphyxo3d3aeh7kn6pguwgv6flnhuhavs5zhh3fi4.cpp(56): note: There is no context in which this conversion is possible
```

Discussed with @jansel , and I change the template code `long` to `int64_t`. it fixed.
image

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @iremyux @Blackhex @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @ezyang @chauhang @penguinwu @voznesenskym @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

### CODE REVIEW

File: torch/_inductor/codecache.py [hunk 1]
```diff
@@ -2181,8 +2181,10 @@ class CppPythonBindingsCodeCache(CppCodeCache):
         static void* (*_torchinductor_pyobject_tensor_data_ptr)(PyObject* obj);
 
         template <typename T> static inline T parse_arg(PyObject* args, size_t n) {
+        #ifndef _WIN32 // Check failed on Windows, skip it.
             static_assert(std::is_pointer<T>::value, "arg type must be pointer or long");
```
Comment (jansel) [inline torch/_inductor/codecache.py:2185]: What is the type of this on Windows?  This code won't work properly if `T` is not a pointer.

I think maybe we need to add another template specialization below.  This is how non-pointer types should be handled.

**Review by jansel — DISMISSED** [review]

File: torch/_inductor/codecache.py (hunk 1 above)
Comment (xuhancn) [reply torch/_inductor/codecache.py:2185]: @After remove skip, the error shows the type is `int64_t`. How to handle this type?

```cmd
cl : Command line warning D9025 : overriding '/openmp' with '/openmp:experimental'
ck2c7cx3jqy7t3wydhqzegtvwbwwdattc6n47mb4h6bupagjo64e.cpp
C:/Users/Xuhan/AppData/Local/Temp/torchinductor_Xuhan/k2/ck2c7cx3jqy7t3wydhqzegtvwbwwdattc6n47mb4h6bupagjo64e.cpp(19): warning C4849: OpenMP 'simdlen' clause ignored in 'simd' directive
C:/Users/Xuhan/AppData/Local/Temp/torchinductor_Xuhan/k2/ck2c7cx3jqy7t3wydhqzegtvwbwwdattc6n47mb4h6bupagjo64e.cpp(55): error C2338: static_assert failed: 'arg type must be pointer or long'
C:/Users/Xuhan/AppData/Local/Temp/torchinductor_Xuhan/k2/ck2c7cx3jqy7t3wydhqzegtvwbwwdattc6n47mb4h6bupagjo64e.cpp(79): note: see reference to function template instantiation 'T parse_arg<int64_t>(PyObject *,size_t)' being compiled
        with
        [
            T=int64_t
        ]


Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
```

Comment (jansel) [reply torch/_inductor/codecache.py:2184]: You could try replacing `long` with `int64_t` below, or add another specialization for Windows.

File: torch/_inductor/codecache.py [hunk 2]
```diff
@@ -2184,12 +2184,23 @@ class CppPythonBindingsCodeCache(CppCodeCache):
...
             return static_cast<T>(_torchinductor_pyobject_tensor_data_ptr(PyTuple_GET_ITEM(args, n)));
         }
+
+        #ifdef _WIN32
```
Comment (jansel) [inline torch/_inductor/codecache.py:2188]: Does the windows version work on linux?  Can we remove the ifdef?  I think long and int64_t are the same type on linux.

**Review by jansel — DISMISSED** [review]

File: torch/_inductor/codecache.py [hunk 3]
```diff
@@ -2184,12 +2184,23 @@ class CppPythonBindingsCodeCache(CppCodeCache):
...
+        template <> inline int64_t parse_arg<int64_t>(PyObject* args, size_t n) {
+            auto result = PyLong_AsSsize_t(PyTuple_GET_ITEM(args, n));
+            if(unlikely(result == -1 && PyErr_Occurred()))
+                throw std::runtime_error("expected int arg");
+            return result;
+        }
+        #else
         template <> inline long parse_arg<long>(PyObject* args, size_t n) {
```
Comment (jansel) [inline torch/_inductor/codecache.py:2196]:
```suggestion
        template <> inline int64_t parse_arg<int64_t>(PyObject* args, size_t n) {
```

If you make this change, can you just remove the ifdef entirely?  This should be a no-op change on linux.

**Review by jansel — CHANGES_REQUESTED** [review]

**Comment by xuhancn** [comment]
Hi @jansel I did some experiment about we discussed use long from caller sets the type. It is not safe on Windows use long because of only Windows long is size 4, and other OS long is size 8. 
Detailed in this PR: https://github.com/pytorch/pytorch/pull/133732
Only on Windows long’s size shorter than other OS. It maybe lost precision. It is unsafe.

I think this PR can passed CI can proof its quality. I‘m insisting my opinion, this PR is right solution, and better than use long type.

CC: @jgong5 any comments for this discussion?

**Review by jansel — CHANGES_REQUESTED** [review]
Let's use int64_t on all platforms.

**Pull request closed without merging**